Replace `MenuPrimitive` with `PopoverPrimitive` in `Dropdown` and `Breadcrumb::Truncation`

alex-ju opened the pull request:
### :pushpin: Summary

This PR replaces the `MenuPrimitive` with `PopoverPrimitive` in `Dropdown` and `Breadcrumb::Truncation` with the main benefits of:
 - aligning the `RichTooltip` and `Dropdown` implementations
 - reducing codebase and abstractions by deprecating the `MenuPrimitive`
 - resolving the current issues we have with the `Dropdown` list positioning on the z-axis (including where placed in a `Modal`, or the `AppHeader`)
 - introducing optional collision detection for the `Dropdown` component

### :hammer_and_wrench: Detailed description

 - Replaced the `MenuPrimitive` with `PopoverPrimitive` in `Dropdown` mapping functionality and preserved the current API
 - Introduced two new optional arguments
   - `@enableCollisionDetection` (default to `false`, to preserve the original behavior; we can switch the default to `true` in the future to benefit more scenarios and align with `RichTooltip`)
   - `@isOpen` (allowing one or multiple dropdowns to be open at render)
 - Replaced the `MenuPrimitive` with `PopoverPrimitive` in `Breadcrumb::Truncation` while preserving the current behavior and API
- Deprecated the `Hds::MenuPrimitive`; after these changes, it is not used anywhere else in our codebase or our consumers' codebases so we want to prevent people from using it

### 👷 Remaining tasks

- [x] test in consumers' codebases
- [x] update `showcase` with new arguments
- [x] update `wesbite` with new arguments

### :link: External links

<!-- Issues, RFC, etc. -->
Jira ticket: [HDSP-3631](https://hashicorp.atlassian.net/browse/HDS-3631)

***

### 👀 Component checklist

- [x] Percy was checked for any visual regression
- [x] A changelog entry was added via [Changesets](https://github.com/changesets/changesets) if needed (see [templates here](https://github.com/hashicorp/design-system/blob/main/wiki/Website-Changelog.md#templates-for-npm-packages))

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.


[HDSP-96]: https://hashicorp.atlassian.net/browse/HDSP-96?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

didoo commented on packages/components/src/components/hds/breadcrumb/truncation.hbs:
```diff
@@ -3,24 +3,25 @@
 <li class="hds-breadcrumb__item hds-breadcrumb__item--is-truncation" ...attributes>
-  <Hds::MenuPrimitive>
-    <:toggle as |t|>
+  <Hds::PopoverPrimitive @enableClickEvents={{true}} as |PP|>
```
the fact that it's so simple is 🤯

didoo commented on packages/components/src/components/hds/dropdown/index.hbs:
```diff
@@ -2,22 +2,26 @@
 }}
-<Hds::MenuPrimitive class={{this.classNames}} @onClose={{@onClose}} ...attributes>
-  <:toggle as |t|>
+<Hds::PopoverPrimitive @isOpen={{@isOpen}} @onClose={{@onClose}} @enableClickEvents={{true}} as |PP|>
```
same 🤯

didoo reviewed (commented):
I haven't looked in detail at the code, because it's in draft and I would imagine there will be some refinements here and there. The overall approach I think is great! The fact that the solution was already there is mind-blowing on one side, and encouraging on the other (all the investment that was made on _getting it right™_ the first time, with the PopoverPrimitive and collision detection has paid back with interests!)

I don't have any objections, it seems a great solution: max result with minimum effort (in code, I mean; I am sure that also for you took quite some time to _get it right™_ with your exploration.

alex-ju commented on packages/components/src/components/hds/menu-primitive/index.ts:
```diff
@@ -36,6 +37,24 @@ export default class MenuPrimitiveComponent extends Component<MenuPrimitiveSigna
+        url: 'https://helios.hashicorp.design/components/menu-primitive?tab=version%20history#460',
+        for: '@hashicorp/design-system-components',
+        since: {
+          enabled: '4.8.0',
```
At this point, it will be 4.9.0, but I'll wait until we get this reviewed and approved before committing to a version.

```suggestion
          enabled: '4.9.0',
```

alex-ju replied:
Right! updated in b8c5e9326950331f904bbf7ad52c6a26c1cde62e

didoo commented on showcase/app/templates/components/dropdown.hbs:
```diff
@@ -65,6 +48,31 @@
 
+  <Shw::Divider @level={{2}} />
+
+  <Shw::Text::H2>Collision detection</Shw::Text::H2>
```
👍

didoo reviewed (commented):
Left a few minor comments. The most important things to do before approving are:
- update the changeset
- update the documentation

Also, do you think you can add a "Status" badge (via frontmatter) and `Version history` tab too, to the components that have been updated? Otherwise I can do it, no problem.

alex-ju commented on packages/components/src/components/hds/dropdown/index.ts:
```diff
@@ -103,9 +128,6 @@ export default class HdsDropdownComponent extends Component<HdsDropdownSignature
     const classes = ['hds-dropdown__content'];
 
-    // add a class based on the @listPosition argument
-    classes.push(`hds-dropdown__content--position-${this.listPosition}`);
```
hah.. we could leave the classes in to avoid a breaking change, and make a note to remove them in the next major

alex-ju replied:
left them in for now 583d9338c62574762bee60ca302dc726d226a2c2
added a comment to remove (and flag) them in the next major release

alex-ju commented on packages/components/src/components/hds/dropdown/toggle/button.ts:
```diff
@@ -37,9 +37,9 @@ export interface HdsDropdownToggleButtonSignature {
     icon?: FlightIconSignature['Args']['name'];
     isFullWidth?: boolean;
     isOpen?: boolean;
-    onClick?: (event: MouseEvent) => void;
```
It was intended to be used only within the Dropdown component, but good shout, I'll have a look across the codebases to check for any potential usage

alex-ju replied:
I looked at instances and it's not used anywhere outside our codebase

alex-ju commented on packages/components/src/components/hds/breadcrumb/truncation.hbs:
```diff
@@ -3,24 +3,25 @@
 <li class="hds-breadcrumb__item hds-breadcrumb__item--is-truncation" ...attributes>
-  <Hds::MenuPrimitive>
-    <:toggle as |t|>
+  <Hds::PopoverPrimitive @enableClickEvents={{true}} as |PP|>
```
Updated PR title and changelog entry in ff0ae8fd8862a68e5acd94cc7f49413c2b9564ae

alex-ju commented on packages/components/src/components/hds/dropdown/types.ts:
```diff
@@ -3,10 +3,22 @@
 }
 export type HdsDropdownPositions = `${HdsDropdownPositionValues}`;
+
+export const HdsDropdownPositionToPlacementValues: Record<
```
Documented in 60631cb5739ed283c5402255fb78c04a6e4a1b5d

alex-ju commented on packages/components/src/components/hds/dropdown/index.ts:
```diff
@@ -7,7 +7,10 @@ import Component from '@glimmer/component';
 import { assert } from '@ember/debug';
 
-import { HdsDropdownPositionValues } from './types.ts';
+import {
```
Similarly, documented in 60631cb5739ed283c5402255fb78c04a6e4a1b5d

alex-ju commented on packages/components/src/components/hds/dropdown/index.ts:
```diff
@@ -26,15 +29,19 @@ import type { HdsDropdownToggleButtonSignature } from './toggle/button';
   Args: MenuPrimitiveSignature['Args'] & {
     height?: string;
     isInline?: boolean;
+    isOpen?: boolean;
```
I looked it up but couldn't find it 🤔

didoo replied:
OK, I'm probably misremembering.

alex-ju commented on packages/components/src/components/hds/menu-primitive/index.hbs:
```diff
@@ -2,6 +2,9 @@
   SPDX-License-Identifier: MPL-2.0
 }}
+{{!
+  THIS COMPONENT IS NOW DEPRECATED
```
Updated in 8f044473da727551f63f900c67c5e66249782a08

alex-ju commented on packages/components/src/styles/components/breadcrumb.scss:
```diff
@@ -181,15 +181,24 @@ $hds-breadcrumb-item-visual-horizontal-padding: 4px;
   box-shadow: var(--token-surface-high-box-shadow);
+
+  // the "popover" attributes comes with pre-defined styling so we need to override it
+  :where(&[popover]) {
```
Agree, I had the same thought – I'll write up a task to try and abstract these https://hashicorp.atlassian.net/browse/HDS-3655

alex-ju commented:
> Left a few minor comments. The most important things to do before approving are:
> 
> * update the changeset
> * update the documentation
> 
> Also, do you think you can add a "Status" badge (via frontmatter) and `Version history` tab too, to the components that have been updated? Otherwise I can do it, no problem.

Thanks for the prompt review on this!
I hope I addressed all the comments. Added the frontmatter and version history in 6291f23b600670de0fda9699b7b2ca87de2231f7

MelSumner commented:
Tested keyboard navigation on here and appreciate that the dropdown is closing when the user keyboard navigates out of it (esp in the modal scenario). 👍
